pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+// storePath: directory for persistent storage
+func NewBLAKE3Store(storePath string) (*BLAKE3Store, error) {
+	// Ensure storage directory exists
+	if err := os.MkdirAll(storePath, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create storage directory: %w", err)
+	}
+
+	store := &BLAKE3Store{
+		storePath:  storePath,
+		cache:      make(map[string][]byte),
+		keyCache:   make(map[string]string),
+		memoryMode: false, // Default to persistent mode
+		writeQueue: make(chan writeOp, 1000), // Buffered channel for async writes
+		errorQueue: make(chan error, 100),    // Buffered channel for errors
+	}
+	
+	// Initialize hasher pool for zero-allocation hashing
+	store.hasherPool = sync.Pool{
+		New: func() interface{} {
+			return blake3.New(32, nil) // Pre-configured 256-bit unkeyed hasher
+		},
+	}
+	
+	// Start background writer goroutine
+	go store.backgroundWriter()
+	
+	// Start error handler goroutine
+	go store.errorHandler()
+	
+	return store, nil
+}
+
+// backgroundWriter handles async disk writes for maximum performance
+func (s *BLAKE3Store) backgroundWriter() {
+	for writeOp := range s.writeQueue {
+		if err := os.WriteFile(writeOp.filePath, writeOp.content, 0644); err != nil {
+			// Send error to error handler for logging/retry
+			select {
+			case s.errorQueue <- fmt.Errorf("background write failed for %s: %w", writeOp.filePath, err):
+			default:
+				// Error queue full, error will be dropped but write continues
+			}
+		}
+		s.wg.Done()
+	}
+}
+
+// errorHandler processes background write errors
+func (s *BLAKE3Store) errorHandler() {
+	for err := range s.errorQueue {
+		// Log error to stderr for proper error handling
+		// Using fmt.Fprintf to stderr instead of stdout for errors
+		fmt.Fprintf(os.Stderr, "[ERROR] BLAKE3Store background write failed: %v\n", err)
+		
+		// TODO: In production, this should:
+		// 1. Log to proper logging system (slog, logrus, zap)
+		// 2. Implement retry logic with exponential backoff
+		// 3. Send metrics/alerts for monitoring
+		// 4. Consider circuit breaker pattern for persistent failures
+	}
+}
+
+// EnableMemoryMode switches to memory-only mode for maximum performance (<70μs targets)
+func (s *BLAKE3Store) EnableMemoryMode() {
+	s.memoryMode = true
+}
+
+// Store saves content and returns its BLAKE3 hash
+func (s *BLAKE3Store) Store(content []byte) (types.Hash, error) {
+	// Quick check if store is closed (avoid lock for performance)
+	if s.closed {
+		return types.Hash{}, fmt.Errorf("store is closed")
+	}
+	// Get hasher from pool for zero allocation
+	hasher := s.hasherPool.Get().(*blake3.Hasher)
+	defer func() {
+		hasher.Reset() // Reset for reuse
+		s.hasherPool.Put(hasher)
+	}()
+
+	// Compute BLAKE3 hash using pooled hasher
+	hasher.Write(content)
+	digest := hasher.Sum(nil)
+
+	hash := types.Hash{
+		Algorithm: types.BLAKE3,
+		Digest:    digest,
+	}
+
+	// Pre-compute hash key once (use hex digest as key)
+	hashKey := fmt.Sprintf("%x", digest)
```

Comment:
Micro-optimization: use encoding/hex (hex.EncodeToString) instead of fmt.Sprintf("%x", ...) on a hot path to reduce allocs/CPU.